Switch to cryptographically strong PRNG

## Problem

In RTP, the SSRC value and the initial value for seqnums and timestamps should be random.

Here is what [RFC 3550](https://tools.ietf.org/html/rfc3550) says about seqnums:

> The initial value of the sequence number
      SHOULD be random (unpredictable) to make known-plaintext attacks
      on encryption more difficult, even if the source itself does not
      encrypt according to the method in Section 9.1, because the
      packets may flow through a translator that does.  Techniques for
      choosing unpredictable numbers are discussed in [17].

These requirements have security reasons. The RFC recommends using a cryptographically strong PRNG (CSPRNG), but currently our PRNG is uniform but not cryptographically strong.

## Solution

We already use libuv, which also provides CSPRNG: http://docs.libuv.org/en/v1.x/misc.html#c.uv_random

Here is what we should do:

* Rename current PRNG: `roc_core/target_posix/roc_core/random.h|cpp` to `roc_core/target_posix/roc_core/fast_random.h|cpp` and `core::random()` to `core::fast_random()`.

* Add `roc_core/target_libuv/roc_core/secure_random.h|cpp` with `core::secure_random()`. Implement is using synchronous version of `uv_random()` (set `loop`, `req`, and `cb` to NULL).

* Use `secure_random()` instead of `fast_random()` in `audio::Packetizer` and `fec::Writer`.

thanks! I met an issue: though I have `#include <uv.h>` in `secure_random.cpp`, I have `src/modules/roc_core/target_posix/roc_core/secure_random.cpp:56:13: error: use of undeclared identifier 'uv_random'
        if (uv_random(NULL, NULL, &val, sizeof(val), 0, NULL) && val >= min) {`

--- a/docs/sphinx/internals/code_structure.rst
+++ b/docs/sphinx/internals/code_structure.rst
@@ -107,8 +107,8 @@ Example directory structure employing targets:
     ├── target_posix
     │   └── roc_core
     │       ├── ...
-    │       ├── random.cpp
-    │       └── random.h
+    │       ├── fast_random.cpp
+    │       └── fast_random.h
     ├── target_posixtime
     │   └── roc_core
     │       ├── ...

--- a/src/modules/roc_audio/packetizer.cpp
+++ b/src/modules/roc_audio/packetizer.cpp
@@ -9,7 +9,7 @@
 #include "roc_audio/packetizer.h"
 #include "roc_core/log.h"
 #include "roc_core/panic.h"
-#include "roc_core/random.h"
+#include "roc_core/secure_random.h"
 
 namespace roc {
 namespace audio {
@@ -35,11 +35,24 @@ Packetizer::Packetizer(packet::IWriter& writer,
     , payload_type_(payload_type)
     , payload_size_(payload_encoder.encoded_size(samples_per_packet_))
     , packet_pos_(0)
-    , source_((packet::source_t)core::random(0, packet::source_t(-1)))
-    , seqnum_((packet::seqnum_t)core::random(0, packet::seqnum_t(-1)))
-    , timestamp_((packet::timestamp_t)core::random(0, packet::timestamp_t(-1))) {
+    , valid_(false) {
+    uint32_t rand_source = 0, rand_seqnum = 0, rand_timestamp = 0;
+    if (!core::secure_random(0, packet::source_t(-1), rand_source)
+        || !core::secure_random(0, packet::seqnum_t(-1), rand_seqnum)
+        || !core::secure_random(0, packet::timestamp_t(-1), rand_timestamp)) {
+        roc_log(LogError, "packetizer: initializing fails");
+        return;
+    }
+    source_ = (packet::source_t)rand_source;
+    seqnum_ = (packet::seqnum_t)rand_seqnum;
+    timestamp_ = (packet::timestamp_t)rand_timestamp;
+    valid_ = true;
     roc_log(LogDebug, "packetizer: initializing: n_channels=%lu samples_per_packet=%lu",
             (unsigned long)num_channels_, (unsigned long)samples_per_packet_);
+}
+
+bool Packetizer::valid() const {
+    return valid_;
 }
 
 void Packetizer::write(Frame& frame) {

--- a/src/modules/roc_audio/packetizer.h
+++ b/src/modules/roc_audio/packetizer.h
@@ -62,6 +62,9 @@ public:
     //!  Packet is padded to match fixed size.
     void flush();
 
+    //! Check if object is successfully constructed.
+    bool valid() const;
+
 private:
     bool begin_packet_();
     void end_packet_();
@@ -86,9 +89,11 @@ private:
     packet::PacketPtr packet_;
     size_t packet_pos_;
 
-    const packet::source_t source_;
+    packet::source_t source_;
     packet::seqnum_t seqnum_;
     packet::timestamp_t timestamp_;
+
+    bool valid_;
 };
 
 } // namespace audio

--- a/src/modules/roc_core/target_posix/roc_core/fast_random.cpp
+++ b/src/modules/roc_core/target_posix/roc_core/fast_random.cpp
@@ -10,8 +10,8 @@
 #include <stdlib.h>
 
 #include "roc_core/errno_to_str.h"
+#include "roc_core/fast_random.h"
 #include "roc_core/panic.h"
-#include "roc_core/random.h"
 #include "roc_core/time.h"
 
 namespace roc {
@@ -35,7 +35,7 @@ unsigned short rand_seed[3] = {};
 // race is harmless. See https://www.evanjones.ca/random-thread-safe.html.
 //
 // This implementation is not a cryptographically strong PRNG.
-uint32_t random(uint32_t from, uint32_t to) {
+uint32_t fast_random(uint32_t from, uint32_t to) {
     roc_panic_if_not(from <= to);
 
     const uint64_t upper = uint64_t(to) - from + 1;

--- a/src/modules/roc_core/target_posix/roc_core/fast_random.h
+++ b/src/modules/roc_core/target_posix/roc_core/fast_random.h
@@ -6,23 +6,23 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
-//! @file roc_core/target_posix/roc_core/random.h
+//! @file roc_core/target_posix/roc_core/fast_random.h
 //! @brief Helpers to work with random numbers.
 
-#ifndef ROC_CORE_RANDOM_H_
-#define ROC_CORE_RANDOM_H_
+#ifndef ROC_CORE_FAST_RANDOM_H_
+#define ROC_CORE_FAST_RANDOM_H_
 
 #include "roc_core/stddefs.h"
 
 namespace roc {
 namespace core {
 
-//! Get random integer.
+//! Get a random integer from a non cryptographically secure, but fast PRNG.
 //! Thread-safe.
 //! @returns random value in range [from; to].
-uint32_t random(uint32_t from, uint32_t to);
+uint32_t fast_random(uint32_t from, uint32_t to);
 
 } // namespace core
 } // namespace roc
 
-#endif // ROC_CORE_RANDOM_H_
+#endif // ROC_CORE_FAST_RANDOM_H_

--- a/src/modules/roc_core/target_posix/roc_core/secure_random.cpp
+++ b/src/modules/roc_core/target_posix/roc_core/secure_random.cpp
@@ -1,0 +1,55 @@
+/*
+ * Copyright (c) 2020 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <uv.h>
+
+#include "roc_core/fast_random.h"
+#include "roc_core/log.h"
+#include "roc_core/panic.h"
+#include "roc_core/secure_random.h"
+
+namespace roc {
+namespace core {
+
+// Based on uv_random from libuv and arc4random_uniform() from OpenBSD.
+#if UV_VERSION_MAJOR > 1 || (UV_VERSION_MAJOR == 1 && UV_VERSION_MINOR >= 33)
+bool secure_random(uint32_t from, uint32_t to, uint32_t& result) {
+    roc_panic_if_not(from <= to);
+
+    const uint64_t upper = (uint64_t)to - from + 1;
+    const uint64_t min = -upper % upper;
+    uint64_t val;
+
+    for (;;) {
+        if (int err = uv_random(NULL, NULL, &val, sizeof(val), 0, NULL)) {
+            roc_log(LogError, "secure random : uv_random(): [%s] %s", uv_err_name(err),
+                    uv_strerror(err));
+            return false;
+        }
+        if (val >= min) {
+            break;
+        }
+    }
+
+    const uint32_t ret = from + (uint32_t)(val % upper);
+
+    roc_panic_if_not(ret >= from);
+    roc_panic_if_not(ret <= to);
+
+    result = ret;
+    return true;
+}
+#else  // UV_VERSION_MAJOR < 1 || (UV_VERSION_MAJOR == 1 && UV_VERSION_MINOR < 33)
+bool secure_random(uint32_t from, uint32_t to, uint32_t& result) {
+    result = fast_random(from, to);
+    return true;
+}
+#endif // UV_VERSION_MAJOR > 1 || (UV_VERSION_MAJOR == 1 && UV_VERSION_MINOR >= 33)
+
+} // namespace core
+} // namespace roc

--- a/src/modules/roc_core/target_posix/roc_core/secure_random.h
+++ b/src/modules/roc_core/target_posix/roc_core/secure_random.h
@@ -1,0 +1,32 @@
+/*
+ * Copyright (c) 2020 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+//! @file roc_core/target_posix/roc_core/secure_random.h
+//! @brief Helpers to work with CSPRNG.
+
+#ifndef ROC_CORE_SECURE_RANDOM_H_
+#define ROC_CORE_SECURE_RANDOM_H_
+
+#include "roc_core/stddefs.h"
+
+namespace roc {
+namespace core {
+
+//! Get a random integer from a cryptographically secure pseudorandom number
+//! generator(CSPRNG).
+//! @remarks
+//!  Writes a random value in [from; to] to @p result.
+//!  May block indefinitely.
+//!  Falls back to a regular PRNG when using libuv version before 1.33.0.
+//! @returns true on success or false if accessing CSPRNG failed.
+bool secure_random(uint32_t from, uint32_t to, uint32_t& result);
+
+} // namespace core
+} // namespace roc
+
+#endif // ROC_CORE_SECURE_RANDOM_H_

--- a/src/modules/roc_fec/writer.cpp
+++ b/src/modules/roc_fec/writer.cpp
@@ -9,7 +9,7 @@
 #include "roc_fec/writer.h"
 #include "roc_core/log.h"
 #include "roc_core/panic.h"
-#include "roc_core/random.h"
+#include "roc_core/secure_random.h"
 #include "roc_packet/fec_scheme_to_str.h"
 
 namespace roc {
@@ -37,12 +37,17 @@ Writer::Writer(const WriterConfig& config,
     , buffer_pool_(buffer_pool)
     , repair_block_(allocator)
     , first_packet_(true)
-    , cur_sbn_((packet::blknum_t)core::random(0, packet::blknum_t(-1)))
-    , cur_block_repair_sn_((packet::seqnum_t)core::random(0, packet::seqnum_t(-1)))
     , cur_packet_(0)
     , fec_scheme_(fec_scheme)
     , valid_(false)
     , alive_(true) {
+    uint32_t rand_cur_sbn = 0, rand_cur_block_repair_sn = 0;
+    if (!core::secure_random(0, packet::blknum_t(-1), rand_cur_sbn)
+        || !core::secure_random(0, packet::seqnum_t(-1), rand_cur_block_repair_sn)) {
+        return;
+    }
+    cur_sbn_ = (packet::blknum_t)rand_cur_sbn;
+    cur_block_repair_sn_ = (packet::seqnum_t)rand_cur_block_repair_sn;
     if (!resize(config.n_source_packets, config.n_repair_packets)) {
         return;
     }

--- a/src/modules/roc_packet/interleaver.cpp
+++ b/src/modules/roc_packet/interleaver.cpp
@@ -7,8 +7,8 @@
  */
 
 #include "roc_packet/interleaver.h"
+#include "roc_core/fast_random.h"
 #include "roc_core/log.h"
-#include "roc_core/random.h"
 
 namespace roc {
 namespace packet {
@@ -81,7 +81,7 @@ void Interleaver::reinit_seq_() {
         send_seq_[i] = i;
     }
     for (size_t i = block_size_; i > 0; --i) {
-        const size_t j = core::random(0, (unsigned int)i - 1);
+        const size_t j = core::fast_random(0, (unsigned int)i - 1);
         const size_t buff = send_seq_[i - 1];
         send_seq_[i - 1] = send_seq_[j];
         send_seq_[j] = buff;

--- a/src/modules/roc_pipeline/sender_endpoint_set.cpp
+++ b/src/modules/roc_pipeline/sender_endpoint_set.cpp
@@ -209,7 +209,7 @@ bool SenderEndpointSet::create_pipeline_() {
                           config_.packet_length, format->sample_rate,
                           config_.payload_type),
                       allocator_);
-    if (!packetizer_) {
+    if (!packetizer_ || !packetizer_->valid()) {
         return false;
     }
 

--- a/src/tests/roc_audio/test_helpers/awgn.h
+++ b/src/tests/roc_audio/test_helpers/awgn.h
@@ -9,7 +9,7 @@
 #ifndef ROC_AUDIO_TEST_HELPERS_AWGN_H_
 #define ROC_AUDIO_TEST_HELPERS_AWGN_H_
 
-#include "roc_core/random.h"
+#include "roc_core/fast_random.h"
 
 namespace roc {
 namespace audio {
@@ -21,8 +21,8 @@ namespace test {
 inline double generate_awgn() {
     enum { RandMax = 2000000000 };
 
-    const double r1 = core::random(0, RandMax) / (double)RandMax;
-    const double r2 = core::random(10, RandMax) / (double)RandMax;
+    const double r1 = core::fast_random(0, RandMax) / (double)RandMax;
+    const double r2 = core::fast_random(10, RandMax) / (double)RandMax;
 
     return sqrt(-2.0 * log(r2)) * cos(2.0 * M_PI * r1);
 }

--- a/src/tests/roc_audio/test_resampler.cpp
+++ b/src/tests/roc_audio/test_resampler.cpp
@@ -18,7 +18,6 @@
 #include "roc_audio/resampler_reader.h"
 #include "roc_core/buffer_pool.h"
 #include "roc_core/heap_allocator.h"
-#include "roc_core/random.h"
 #include "roc_core/scoped_ptr.h"
 #include "roc_core/stddefs.h"
 

--- a/src/tests/roc_fec/test_encoder_decoder.cpp
+++ b/src/tests/roc_fec/test_encoder_decoder.cpp
@@ -10,9 +10,9 @@
 
 #include "roc_core/array.h"
 #include "roc_core/buffer_pool.h"
+#include "roc_core/fast_random.h"
 #include "roc_core/heap_allocator.h"
 #include "roc_core/log.h"
-#include "roc_core/random.h"
 #include "roc_core/scoped_ptr.h"
 #include "roc_fec/codec_map.h"
 
@@ -86,7 +86,7 @@ private:
         core::Slice<uint8_t> buf = new (buffer_pool) core::Buffer<uint8_t>(buffer_pool);
         buf.resize(p_size);
         for (size_t j = 0; j < buf.size(); ++j) {
-            buf.data()[j] = (uint8_t)core::random(0, 0xff);
+            buf.data()[j] = (uint8_t)core::fast_random(0, 0xff);
         }
         return buf;
     }
@@ -172,7 +172,7 @@ TEST(encoder_decoder, random_losses) {
 
             size_t curr_loss = 0;
             for (size_t i = 0; i < NumSourcePackets + NumRepairPackets; ++i) {
-                if (core::random(0, 100) < LossPercent && curr_loss <= MaxLoss) {
+                if (core::fast_random(0, 100) < LossPercent && curr_loss <= MaxLoss) {
                     total_loss++;
                     curr_loss++;
                 } else {

